### DISCUSSION TITLE
How flexible are record keys/PDS paths?

### DISCUSSION THREAD

**Discussion by snarfed** [bluesky-social/atproto#844] [Clarifications and "Why?"] [answered]
Hey Bluesky team! Are you all parsing or otherwise validating record key paths against any specific format? https://atproto.com/guides/data-repos#identifier-types says they should be based on TIDs, ~which https://staging.bsky.app/ post URLs mostly seem to follow, minus the dashes~ (never mind, web frontend URLs obviously aren't the same as PDS `at://` URLs). Will federation depend on rkey format though? ie if I wanted to use an entirely different rkey path scheme, would that be ok? Looking at the current lexicons, some of the `app.bsky` data defs have explicit `tid` fields, which is fine, but everything in `com.atproto` seems to use CIDs and full rkeys, no explicit TIDs anywhere.

Thanks in advance!

Somewhat related:

- https://github.com/bluesky-social/atproto/discussions/114#discussioncomment-2704155
- https://github.com/bluesky-social/atproto/issues/334
- https://github.com/bluesky-social/atproto/issues/411

**Comment by dholms** [comment, marked as answer]
Rkey is left intentionally flexible. You can see in many of the record lexicons, they declare their key type. We only have two supported key types right now:
- `tid` (for most collections)
- `literal:self` (for collections that should only have one record in them)

Additional rkey semantics are possible & supported, however I'd encourage you to think twice before doing so. Do you have anything in particular in mind?

TIDs are useful because they order collection contents by rough creation time. When syncing some slice of a repository, it is likely that you'll be doing so based on time & this allows creates smaller aggregate proofs.

An example of this is follows. Wouldn't it make sense to key them by did of the account that you're following?
In some sense yes. Especially if the repo is being used primarily for reads. But we view the repo as primarily a sync utility. If I add 10 new follows & use TIDs as keys, then all 10 follows will share much of their proof chain. Syncing is there for easier - likely half (or less!) of the bandwidth. If we key them by DID, then they will be uniformly scattered throughout the collection & share much less of a proof path.

**Reply by dholms** [reply]
Oh I should add that pathing in general _is_ enforced.

MST paths must be of the form `${nsid}/${rkey}` where `nsid` is a valid nsid (such as `app.bsky.feed.post`) & `rkey` is some valid rkey (which is loosely defined right now, but it only supports ascii & is limited in length)

**Comment by snarfed** [comment]
Thanks! This is very helpful.

> Do you have anything in particular in mind?

Yes! I'm currently working on adding Bluesky federation (jumping the gun, I know, apologies!) to https://fed.brid.gy/, which is a bridge to/from IndieWeb sites. This means it won't be a standard PDS; instead, it's adapting posts and other user interactions from their existing authoritative sources elsewhere, specifically users' web sites. That means those posts already have canonical ids: their http URLs. I can definitely mint TIDs and make compatible rkeys and store the mapping, and that layer of indirection may be useful regardless, but I was just curious whether I _could_ drop in their full (encoded) URLs as rkeys.

Sounds like the answer is...not right now, and possible but not likely in the future. Totally ok!

**Reply by dholms** [reply]
Ah yup i think that's an anti-pattern. Everything in the repo is sort-ordered & the general rule of thumb for datakeys is:
- records should live next to similar records (ie posts next to posts, likes next to likes)
- within those buckets, records should live next to records that were created around the same time

Also full-encoded urls are probably too large. I'll have to doublecheck what our allowance is for rkey size, but i wanna say it's 32 ascii chars

The network relies on secondary indexes & I think you can get most of what you like by adding an extra key to the record that includes the canonical url in the other network!

Thanks for checking in on this! We need to get some docs together with best practices like these

**Reply by bnewbold** [reply]
to clarify, @dholms, when you mention adding an extra key to the record, do you mean on `app.bsky.feed.post` records, or having @snarfed create their own new record type and then they can do whatever they want?

**Comment by bnewbold** [comment]
This is pretty well covered in the specs now: https://atproto.com/specs/record-key

Going to close as outdated, but feel free to re-open if you disagree... or if you think the spec should be different.
